src/connections.jl

```diff
@@ -128,6 +149,39 @@ function handle_new_connection(jl_conn::Connection; throw_error::Bool=true)
     return jl_conn
 end
 
+# Perform the connection loop as specified in
+# https://www.postgresql.org/docs/10/libpq-connect.html#LIBPQ-PQCONNECTSTARTPARAMS
+# NOTE: this uses the non-blocking interface but this function does block!
+function _connect_nonblocking(keywords, values, expand_dbname)
+    conn = libpq_c.PQconnectStartParams(keywords, values, expand_dbname)
+    c_state = libpq_c.PQstatus(conn)
+
+    # This is also true when `conn == C_NULL`
+    c_state == libpq_c.CONNECTION_BAD && return conn
+
+    p_state = libpq_c.PGRES_POLLING_WRITING
+    while p_state != libpq_c.PGRES_POLLING_FAILED && p_state != libpq_c.PGRES_POLLING_OK
+        # PostgreSQL docs say we can't assume that the socket will remain the same while
+        # connecting
+        if p_state == libpq_c.PGRES_POLLING_WRITING
+            wait(socket(conn); writable=true)
+        elseif p_state == libpq_c.PGRES_POLLING_READING
+            wait(socket(conn); readable=true)
+        end
+
+        p_state = libpq_c.PQconnectPoll(conn)
+
+        # Only print connection status when it changes
+        new_c_state = libpq_c.PQstatus(conn)
+        if new_c_state != c_state
+            c_state = new_c_state
+            debug(LOGGER, "Connection $conn: $(CONNECTION_STATUS_MESSAGES[c_state])")
```

Comment:
I'd probably just use a regular `Dict` and do:
```suggestion
            msg = get(CONNECTION_STATUS_MESSAGES, c_state, "Connecting")
            debug(LOGGER, "Connection $conn: $msg")
```

Reply:
Yes that's better